### PR TITLE
Introduce "pre-compiled" terms for buy gas, coinbase

### PR DESCRIPTION
**Pull request by slpopejoy**
Gets a ~1ms savings per tx, and ~3ms per block

Before:
```
benchmarking PactService/block-new-valid[1]
time                 264.2 ms   (262.9 ms .. 265.3 ms)
                     
benchmarking PactService/block-new-valid[10]
time                 309.0 ms   (296.8 ms .. 325.9 ms)
                     
benchmarking PactService/block-new-valid[50]
time                 506.9 ms   (497.7 ms .. 514.7 ms)
                     
benchmarking PactService/block-new-valid[100]
time                 742.6 ms   (705.6 ms .. 778.2 ms)
                     
benchmarking PactService/block-new[1]
time                 134.1 ms   (132.9 ms .. 135.8 ms)
                     
benchmarking PactService/block-new[10]
time                 158.8 ms   (156.8 ms .. 163.9 ms)
                     
benchmarking PactService/block-new[50]
time                 247.4 ms   (242.8 ms .. 251.0 ms)
                     
benchmarking PactService/block-new[100]
time                 370.4 ms   (359.7 ms .. 388.2 ms)
```

After:
```
benchmarking PactService/block-new-valid[1]
time                 248.6 ms   (233.6 ms .. 269.9 ms)
                      
benchmarking PactService/block-new-valid[10]
time                 288.6 ms   (267.9 ms .. 308.2 ms)
                      
benchmarking PactService/block-new-valid[50]
time                 483.7 ms   (455.1 ms .. 550.2 ms)
                      
benchmarking PactService/block-new-valid[100]
time                 678.7 ms   (605.1 ms .. 763.6 ms)
                      
benchmarking PactService/block-new[1]
time                 117.7 ms   (113.3 ms .. 125.8 ms)
                      
benchmarking PactService/block-new[10]
time                 135.4 ms   (133.3 ms .. 139.7 ms)
                      
benchmarking PactService/block-new[50]
time                 246.8 ms   (230.6 ms .. 265.7 ms)
                      
benchmarking PactService/block-new[100]
time                 378.9 ms   (352.1 ms .. 438.9 ms)
```

### CODE REVIEW

File: src/Chainweb/Pact/Templates.hs [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+buyGasTemplate =
+  (app (qn "coin" "fund-tx")
+   [strLit "sender"
+   ,strLit "mid"
+   ,app (bn "read-keyset") [strLit "miner-keyset"]
+   ,app (bn "read-decimal") [strLit "total"]
+   ]
+  ,strArgSetter 0
+  ,strArgSetter 1
+  )
```
Comment (emilypi) [inline src/Chainweb/Pact/Templates.hs]: NOINLINE here

File: src/Chainweb/Pact/Templates.hs [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+
+dummyParsedCode :: ParsedCode
+dummyParsedCode = ParsedCode "1" [ELiteral $ LiteralExp (LInteger 1) def]
```
Comment (emilypi) [inline src/Chainweb/Pact/Templates.hs]: NOINLINE

File: src/Chainweb/Pact/Templates.hs [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+{-# INLINABLE mkBuyGasTerm #-}
+
+
+coinbaseTemplate :: (Term Name,ASetter' (Term Name) Text)
+coinbaseTemplate =
+  (app (qn "coin" "coinbase")
+   [strLit "mid"
+   ,app (bn "read-keyset") [strLit "miner-keyset"]
+   ,app (bn "read-decimal") [strLit "reward"]]
+  ,strArgSetter 0)
```
Comment (emilypi) [inline src/Chainweb/Pact/Templates.hs]: NOINLINE

File: src/Chainweb/Pact/Templates.hs [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+import Pact.Parse
+import Pact.Types.Command
+import Pact.Types.RPC
+import Pact.Types.Runtime
+
+import Chainweb.Miner.Pact
+import Chainweb.Pact.Types
+
+app :: Name -> [Term Name] -> Term Name
+app f as = TApp (App (TVar f def) as def) def
```
Comment (emilypi) [inline src/Chainweb/Pact/Templates.hs]: we can inline `app` `qn`, `bn` and `strLit`

Comment (slpopejoy) [reply src/Chainweb/Pact/Templates.hs]: INLINABLE?

Comment (emilypi) [reply src/Chainweb/Pact/Templates.hs]: `INLINE` would be fine for these

**Comment by slpopejoy** [comment]
solid!
```
benchmarking PactService/block-new-valid[1]
time                 233.6 ms   (228.5 ms .. 237.9 ms)
                     1.000 R²   (0.999 R² .. 1.000 R²)
mean                 229.5 ms   (225.4 ms .. 232.1 ms)
std dev              4.239 ms   (3.011 ms .. 5.524 ms)
variance introduced by outliers: 14% (moderately inflated)

benchmarking PactService/block-new-valid[10]
time                 275.2 ms   (266.1 ms .. 282.8 ms)
                     1.000 R²   (0.999 R² .. 1.000 R²)
mean                 270.0 ms   (268.0 ms .. 272.6 ms)
std dev              3.223 ms   (2.591 ms .. 3.849 ms)
variance introduced by outliers: 16% (moderately inflated)

benchmarking PactService/block-new-valid[50]
time                 451.7 ms   (441.4 ms .. 464.2 ms)
                     1.000 R²   (1.000 R² .. NaN R²)
mean                 456.9 ms   (454.3 ms .. 459.5 ms)
std dev              3.248 ms   (209.6 μs .. 3.987 ms)
variance introduced by outliers: 19% (moderately inflated)

benchmarking PactService/block-new-valid[100]
time                 698.3 ms   (686.6 ms .. 705.7 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 695.1 ms   (691.7 ms .. 697.3 ms)
std dev              3.410 ms   (1.602 ms .. 4.763 ms)
variance introduced by outliers: 19% (moderately inflated)

benchmarking PactService/block-new[1]
time                 113.6 ms   (109.1 ms .. 117.2 ms)
                     0.998 R²   (0.997 R² .. 1.000 R²)
mean                 116.6 ms   (114.5 ms .. 120.3 ms)
std dev              4.599 ms   (1.031 ms .. 6.932 ms)
variance introduced by outliers: 11% (moderately inflated)

benchmarking PactService/block-new[10]
time                 135.4 ms   (133.7 ms .. 137.4 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 134.9 ms   (134.5 ms .. 135.6 ms)
std dev              840.5 μs   (594.4 μs .. 1.173 ms)
variance introduced by outliers: 11% (moderately inflated)

benchmarking PactService/block-new[50]
time                 231.4 ms   (228.7 ms .. 233.2 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 232.0 ms   (231.3 ms .. 233.2 ms)
std dev              1.297 ms   (634.6 μs .. 1.842 ms)
variance introduced by outliers: 14% (moderately inflated)

benchmarking PactService/block-new[100]
time                 353.6 ms   (350.2 ms .. 359.8 ms)
                     1.000 R²   (NaN R² .. 1.000 R²)
mean                 352.6 ms   (351.7 ms .. 353.5 ms)
std dev              1.042 ms   (886.4 μs .. 1.179 ms)
variance introduced by outliers: 19% (moderately inflated)

```

**Comment by slpopejoy** [comment]
Solid improvement!
```
benchmarking PactService/block-new-valid[1]
time                 233.6 ms   (228.5 ms .. 237.9 ms)

benchmarking PactService/block-new-valid[10]
time                 275.2 ms   (266.1 ms .. 282.8 ms)

benchmarking PactService/block-new-valid[50]
time                 451.7 ms   (441.4 ms .. 464.2 ms)

benchmarking PactService/block-new-valid[100]
time                 698.3 ms   (686.6 ms .. 705.7 ms)

benchmarking PactService/block-new[1]
time                 113.6 ms   (109.1 ms .. 117.2 ms)

benchmarking PactService/block-new[10]
time                 135.4 ms   (133.7 ms .. 137.4 ms)

benchmarking PactService/block-new[50]
time                 231.4 ms   (228.7 ms .. 233.2 ms)

benchmarking PactService/block-new[100]
time                 353.6 ms   (350.2 ms .. 359.8 ms)

```

**Pull request merged**